### PR TITLE
OCPBUGS-781: Update the deprecated field APIServerInternalIP to APIServerInternalIPs

### PR DESCRIPTION
**Pull request by hroyrh**
In provisioning/utils.go we reference .PlatformStatus.BareMetal.APIServerInternalIP attribute from the config API but a recent [change](https://github.com/openshift/api/commit/51f399230d604fa013c2bb341040c4ad126e7309) deprecated this field in favour of .APIServerInternalIPs (note plural), this was done to better suit dual stack case.

### CODE REVIEW

**Comment by rhjanders** [comment]
/lgtm
/cc dtantsur
@dtantsur given this is a continuation of METAL-256, I figured it may be worth your look, LGTM.

File: provisioning/utils.go
Comment (dtantsur) [inline provisioning/utils.go:66]: We used "" as an empty value for the `string` type. The slice type has an empty value as well - `nil`, so let's return it. It's especially true for the error case.

**Comment by dtantsur** [comment]
In addition to one comment inline: how was the vendored file updated? I'm surprised to see neither go.mod nor vendor/modules.txt changes.

**Review by dtantsur — CHANGES_REQUESTED** [review]
Yep, confirmed locally: running `go mod vendor` reverts your change. You cannot just copy updated files from upstream into `vendor`, you need to update `go.mod`.

**Comment by rhjanders** [comment]
Thanks for your insights Dmitry. Removing LGTM for now, will re-review once the new revision is up.
/cancel lgtm

File: provisioning/utils.go
Comment (hroyrh) [reply provisioning/utils.go:66]: In case of an error should we fall back to Pod IP ( as is the case for NonePlatform ) or simply return ?

**Comment by rhjanders** [comment]
Thank you for the revised patch @hroyrh.

I had a look at the issue with the updated dependency, I think an approach like this may help (need to add context as an argument in several places in the code, this is just an example):

```
[janders@carbon cluster-baremetal-operator]$ git diff provisioning/baremetal_pod.go
diff --git a/provisioning/baremetal_pod.go b/provisioning/baremetal_pod.go
index 9c3a92f3..ef10c436 100644
--- a/provisioning/baremetal_pod.go
+++ b/provisioning/baremetal_pod.go
@@ -961,7 +961,7 @@ func getMetal3DeploymentSelector(client appsclientv1.DeploymentsGetter, targetNa
        return nil, err
 }
 
-func EnsureMetal3Deployment(info *ProvisioningInfo) (updated bool, err error) {
+func EnsureMetal3Deployment(ctx context.Context, info *ProvisioningInfo) (updated bool, err error) {
        // Create metal3 deployment object based on current baremetal configuration
        // It will be created with the cboOwnedAnnotation
 
@@ -975,7 +975,7 @@ func EnsureMetal3Deployment(info *ProvisioningInfo) (updated bool, err error) {
        }
 
        deploymentRolloutStartTime = time.Now()
-       deployment, updated, err := resourceapply.ApplyDeployment(info.Client.AppsV1(),
+       deployment, updated, err := resourceapply.ApplyDeployment(ctx, info.Client.AppsV1(),
                info.EventRecorder, metal3Deployment, expectedGeneration)
        if err != nil {
                err = fmt.Errorf("unable to apply Metal3 deployment: %w", err)
[janders@carbon cluster-baremetal-operator]$ 
```

**Comment by hroyrh** [comment]
@rhjanders @dtantsur I've merged changes from #305 and made changes accordingly. Can you please take a look again if things look okay. Also, should I first do a rebase first instead of the merge ( to keep the commit history clean ? ) ?

**Comment by dtantsur** [comment]
@hroyrh I'd really appreciate a rebase on top of master to avoid reviewing unrelated changes

**Comment by hroyrh** [comment]
@dtantsur I have rebased against my initial changes, can you take a look ?

File: vendor/github.com/openshift/api/config/v1/types_infrastructure.go
Comment (honza) [inline vendor/github.com/openshift/api/config/v1/types_infrastructure.go:524]: Changes to this file need to accompanied by changes to `go.mod`, and `go.sum`

File: provisioning/utils.go
Comment (honza) [inline provisioning/utils.go:67]: Consider changing the name to `getServerInternalIPs`

File: provisioning/utils.go
Comment (honza) [inline provisioning/utils.go:101]: Let's handle the error here

File: provisioning/utils.go
Comment (honza) [inline provisioning/utils.go:103]: The naked `return` will return whatever is stored in `err` as the error value.

Comment (hroyrh) [reply provisioning/utils.go:103]: @honza should I just print the error using `fmt` or return an error message. And I think I missed the case of `internalIPs=nil`, so will handle it as well.

Comment (honza) [reply provisioning/utils.go:103]: I think you should be able to assign to `err`, and then do a naked return.

```go
internalIPs, err := getServerInternalIPs(osclient)
if err != nil {
    return
}
```

Comment (hroyrh) [reply provisioning/utils.go:106]: @honza does this look okay now ?

File: provisioning/utils.go
Comment (dtantsur) [inline provisioning/utils.go:85]: These 3 should be handled the same way as None, not removed

File: provisioning/utils.go
Comment (dtantsur) [inline provisioning/utils.go:120]: nit: I don't think this change is needed

Comment (hroyrh) [reply provisioning/utils.go:126]: When I do that, the unit tests throw error : `naked return in func "GetIronicIP" with 32 lines of code (nakedret)` , and explicit return also improves readability I think.

File: provisioning/utils.go
Comment (hroyrh) [reply provisioning/utils.go:85]: Done

**Comment by dtantsur** [comment]
/approve

Please check why the CI is failing

**Comment by hroyrh** [comment]
@dtantsur is the sdn-upgrade job also required to pass before merging this pr ?

**Comment by honza** [comment]
Could use a rebase, not a fan of "minor fixes" commits

**Comment by hroyrh** [comment]
> Could use a rebase, not a fan of "minor fixes" commits

Done

**Pull request merged**
